src/lib/storage/dictionary_segment.cpp

```diff
@@ -0,0 +1,133 @@
+#include <unordered_set>
+
+#include "dictionary_segment.hpp"
+#include "type_cast.hpp"
+
+#include "utils/assert.hpp"
+
+namespace opossum {
+
+template <typename T>
+DictionarySegment<T>::DictionarySegment(const std::shared_ptr<AbstractSegment>& abstract_segment) {
+  for (auto index = ChunkOffset{0}; index < abstract_segment->size(); index++) {
+    const auto& value = type_cast<T>(abstract_segment->operator[](index));
+    _dictionary.push_back(value);
+  }
+
+  // use standard library functions to remove duplicate values from dictionary
+  // sorted dictionary also allows us to be more efficient in element lookup
+  std::sort(begin(_dictionary), end(_dictionary));
+  const auto& end_it = std::unique(begin(_dictionary), end(_dictionary));
+  _dictionary.erase(end_it, _dictionary.end());
+  _dictionary.shrink_to_fit();
+
+  // determine how many bits are needed to encode based on dictionary size
+  const auto num_bits = std::ceil(std::log2(_dictionary.size()));
```

Comment:
You could save these function calls by comparing the size against `std::numeric_limit<uint8_t>::max()`